### DISCUSSION TITLE
Deploy EAR fails with class file has wrong version 55.0, should be 52.0

### DISCUSSION THREAD

**Discussion by rso61** [Keyfactor/ejbca-ce#712] [Q&A] [answered]
Hi,
I am new to EJBCA and came across the error below. Google says its a Java Version problem,
Version 55 == Java 11
Version 52 == Java 8
But I tried 8 and 11 (also17 and 21) making sure java and javac have the same version. But the error stil remains the same.
any suggestions?
Thanks Roland

     [echo]
    [javac] /home/user01/ejbca-ce-main/modules/cesecore-common/src/org/cesecore/audit/Auditable.java:28: error: cannot access CryptoToken
    [javac] import com.keyfactor.util.keys.token.CryptoToken;
    [javac]                                     ^
    [javac]   bad class file: /home/user01/ejbca-ce-main/lib/x509-common-util-3.2.0.jar(com/keyfactor/util/keys/token/CryptoToken.class)
    [javac]     class file has wrong version 55.0, should be 52.0
    [javac]     Please remove or make sure it appears in the correct subdirectory of the classpath.

**Comment by primetomas** [comment]
version 52 means Java 8, so this means that you try to build EJBCA with Java 8, while Java 11 is the minimum.
If you have multiple JDKs installed you really have to make sure the rigth one is used. I.e. on Ubuntu for example you need to use "sudo update-java-alternatives".

If you switch java, also always do "ant clean" to remove old leftovers.

**Reply by rso61** [reply]
Thank you Tomas,
I am using Java 11

[user01@ ejbca-ce-main]$ java -version
openjdk version "11.0.25" 2024-10-15 LTS
OpenJDK Runtime Environment (Red_Hat-11.0.25.0.9-1) (build 11.0.25+9-LTS)
OpenJDK 64-Bit Server VM (Red_Hat-11.0.25.0.9-1) (build 11.0.25+9-LTS, mixed mode, sharing)
[user01@ ejbca-ce-main]$ javac -version
javac 11.0.25

and I calling: ant clean deployear

But the error still remains.

Update: I double checked the alternatives: both java and javac are correct

**Comment by primetomas** [comment]
Can you print the first "display-properties" part that shows up when doing "ant build".

**Reply by rso61** [reply]
you probably looking for this one
     [echo] java.version(ant.java)   = 1.8.0_432 (1.8)
did not recognized it before ...  and where does it come from?

**Reply by primetomas** [reply]
have you run "update-java-alternatives" to switch java implementation or do you just update that local path?

**Reply by rso61** [reply]
under RHEL ist "alternatives" and it looks like this ...

[user01@ejbca-ce-main]$ sudo alternatives --config java
 Selection    Command
-----------------------------------------------
   1           java-17-openjdk.x86_64 (/usr/lib/jvm/java-17-openjdk-17.0.13.0.11-3.el8.x86_64/bin/java)
*  2           java-1.8.0-openjdk.x86_64 (/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.432.b06-2.el8.x86_64/jre/bin/java)
   3           java-21-openjdk.x86_64 (/usr/lib/jvm/java-21-openjdk-21.0.5.0.10-3.el8.x86_64/bin/java)
   4           java-latest-openjdk.x86_64 (/usr/lib/jvm/java-23-openjdk-23.0.0.0.37-1.rolling.el8.x86_64/bin/java)
 + 5           java-11-openjdk.x86_64 (/usr/lib/jvm/java-11-openjdk-11.0.25.0.9-2.el8.x86_64/bin/java)

so it means even I have run alternatives and select Java11, Java8 ist still active. :-O

I will try to remove Java8 ... keep you posted

**Reply by rso61** [reply]
Even when Java8 is no longer in the alternatives list
  Selection    Command
-----------------------------------------------
*+ 1           java-11-openjdk.x86_64 (/usr/lib/jvm/java-11-openjdk-11.0.25.0.9-2.el8.x86_64/bin/javac)
   2           java-17-openjdk.x86_64 (/usr/lib/jvm/java-17-openjdk-17.0.13.0.11-3.el8.x86_64/bin/javac)
   3           java-21-openjdk.x86_64 (/usr/lib/jvm/java-21-openjdk-21.0.5.0.10-3.el8.x86_64/bin/javac)
   4           java-latest-openjdk.x86_64 (/usr/lib/jvm/java-23-openjdk-23.0.0.0.37-1.rolling.el8.x86_64/bin/javac)

property is still set
     [echo] java.version(ant.java)   = 1.8.0_432 (1.8)

and error remains

I'll keep on searching where this is being set ... keep you posted

**Comment by rso61** [comment, marked as answer]
after setting JAVA_HOME explicitly
JAVA_HOME=/usr/lib/jvm/java-11-openjdk-11.0.25.0.9-2.el8.x86_64
it works as expected.
Thanky Tomas for helping me!
